File: Docker/meridian/meridian/crypto.py
"""Crypto primitives for the Meridian server.

Implements the wire-protocol key exchange and AEAD envelope described in
docs/protocol.md:

    shared = X25519(server_priv, client_pub)
    key    = HKDF-SHA256(ikm=shared, salt=client_nonce||server_nonce, info="meridian-v1")
    ct     = AES-256-GCM(key, nonce, aad="meridian/v1/<session_id>")(plaintext)
"""

from __future__ import annotations

import base64
import os

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey,
    X25519PublicKey,
)
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

from . import AAD_INFO

NONCE_LEN = 12
SALT_LEN = 16
KEY_LEN = 32

#: Compact (binary) framing tags for low-bandwidth transports (DNS).
FRAME_KEX = 0x01
FRAME_CHECKIN = 0x02


def b64url_e(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode().rstrip("=")


def b64url_d(data: str) -> bytes:
    pad = "=" * (-len(data) % 4)
    try:
        return base64.urlsafe_b64decode(data + pad)
    except Exception as exc:
        raise CryptoError("invalid base64url") from exc


class CryptoError(Exception):
    """Raised on any cryptographic failure (bad key, tamper, replay)."""


def b64e(data: bytes) -> str:
    return base64.b64encode(data).decode()


def b64d(data: str) -> bytes:
    try:
        return base64.b64decode(data, validate=True)
    except Exception as exc:  # binascii.Error, ValueError
        raise CryptoError("invalid base64") from exc


def new_salt() -> bytes:
    return os.urandom(SALT_LEN)


class SessionCrypto:
    """Holds one session's key material and envelope (de)serialization."""

    def __init__(self, session_id: str, key: bytes) -> None:
        self.session_id = session_id
        self.key = key
        self._aad = f"{AAD_INFO}/{session_id}".encode()

    @classmethod
    def from_exchange(
        cls,
        session_id: str,
        server_priv: X25519PrivateKey,
        client_pub_b64: str,
        client_nonce_b64: str,
        server_nonce_b64: str,
    ) -> SessionCrypto:
        try:
            client_pub = X25519PublicKey.from_public_bytes(b64d(client_pub_b64))
            salt = b64d(client_nonce_b64) + b64d(server_nonce_b64)
        except CryptoError:
            raise
        except Exception as exc:
            raise CryptoError("bad key material") from exc
        if len(salt) != SALT_LEN * 2:
            raise CryptoError("bad nonce length")
        shared = server_priv.exchange(client_pub)
        key = HKDF(
            algorithm=hashes.SHA256(),
            length=KEY_LEN,
            salt=salt,
            info=AAD_INFO.encode(),
        ).derive(shared)
        return cls(session_id, key)

    def seal(self, plaintext: bytes) -> dict:
        nonce = os.urandom(NONCE_LEN)
        ct = AESGCM(self.key).encrypt(nonce, plaintext, self._aad)
        return {"nonce": b64e(nonce), "ct": b64e(ct)}

    def open(self, envelope: dict) -> bytes:
        try:
            nonce = b64d(envelope["nonce"])
            ct = b64d(envelope["ct"])
        except CryptoError:
            raise
        except Exception as exc:
            raise CryptoError("malformed envelope") from exc
        if len(nonce) != NONCE_LEN:
            raise CryptoError("bad nonce length")
        try:
            return AESGCM(self.key).decrypt(nonce, ct, self._aad)
        except Exception as exc:
            raise CryptoError("decryption failed") from exc

    # ------------------------------------------------ compact binary framing
    def seal_compact(self, plaintext: bytes) -> bytes:
        """Return FRAME_CHECKIN | nonce | AES-GCM ct with AAD = meridian/v1/sid."""
        nonce = os.urandom(NONCE_LEN)
        ct = AESGCM(self.key).encrypt(nonce, plaintext, self._aad)
        return bytes([FRAME_CHECKIN]) + nonce + ct

    def open_compact(self, framed: bytes) -> bytes:
        if len(framed) < 1 + NONCE_LEN + 16 or framed[0] != FRAME_CHECKIN:
            raise CryptoError("bad compact frame")
        nonce = framed[1 : 1 + NONCE_LEN]
        ct = framed[1 + NONCE_LEN :]
        try:
            return AESGCM(self.key).decrypt(nonce, ct, self._aad)
        except Exception as exc:
            raise CryptoError("decryption failed") from exc


def generate_server_keypair() -> tuple[str, str]:
    """Return (private_b64, public_b64) for a fresh server identity."""
    priv = X25519PrivateKey.generate()
    pub = priv.public_key().public_bytes_raw()
    return b64e(priv.private_bytes_raw()), b64e(pub)

File: Docker/meridian/implant/internal/mcrypto/crypto.go
// Package mcrypto implements the Meridian session crypto primitives.
//
// Mirrors meridian/crypto.py: shared secret via X25519, key derived with
// HKDF-SHA256 (salt = client_nonce||server_nonce, info = "meridian-v1") and
// an AES-256-GCM AEAD envelope whose AAD binds every message to the session id.
// Standard library only.
package mcrypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/ecdh"
	"crypto/hkdf"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base32"
	"encoding/base64"
	"errors"
	"hash"
)

const (
	AADInfo      = "meridian/v1"
	NonceLen     = 12
	SaltLen      = 16
	KeyLen       = 32
	FrameKex     = 0x01
	FrameCheckin = 0x02
)

var (
	ErrKey       = errors.New("mcrypto: bad key material")
	ErrFrame     = errors.New("mcrypto: bad compact frame")
	ErrDecrypt   = errors.New("mcrypto: decryption failed")
	ErrLen       = errors.New("mcrypto: bad nonce length")
	ErrBadPubLen = errors.New("mcrypto: server public key must be 32 bytes")
)

// B64 / B64D are standard (RFC 4648, padded) base64, matching the server's b64e/b64d.
func B64(data []byte) string        { return base64.StdEncoding.EncodeToString(data) }
func B64D(s string) ([]byte, error) { return base64.StdEncoding.DecodeString(s) }

// B32E is UPPERCASE base32 without padding, matching dns_listener._b32e.
func B32E(data []byte) string {
	return base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(data)
}

// B32D is the inverse of B32E.
func B32D(s string) ([]byte, error) {
	return base32.StdEncoding.WithPadding(base32.NoPadding).DecodeString(s)
}

// DeriveKey computes the 32-byte session key from the client keypair, the
// server public key and both nonces.
func DeriveKey(clientPriv *ecdh.PrivateKey, serverPubB64, clientNonceB64, serverNonceB64 string) ([]byte, error) {
	serverPubRaw, err := B64D(serverPubB64)
	if err != nil {
		return nil, err
	}
	if len(serverPubRaw) != 32 {
		return nil, ErrBadPubLen
	}
	serverPub, err := ecdh.X25519().NewPublicKey(serverPubRaw)
	if err != nil {
		return nil, ErrKey
	}
	shared, err := clientPriv.ECDH(serverPub)
	if err != nil {
		return nil, ErrKey
	}
	clientNonce, err := B64D(clientNonceB64)
	if err != nil {
		return nil, err
	}
	serverNonce, err := B64D(serverNonceB64)
	if err != nil {
		return nil, err
	}
	salt := make([]byte, 0, len(clientNonce)+len(serverNonce))
	salt = append(salt, clientNonce...)
	salt = append(salt, serverNonce...)
	if len(salt) != SaltLen*2 {
		return nil, ErrLen
	}
	key, err := hkdf.Key[hash.Hash](sha256.New, shared, salt, AADInfo, KeyLen)
	if err != nil {
		return nil, err
	}
	return key, nil
}

// Session holds one session's key material.
type Session struct {
	SessionID string
	key       []byte
}

// New builds a session crypto from a derived key.
func New(sessionID string, key []byte) (*Session, error) {
	if len(key) != KeyLen {
		return nil, ErrKey
	}
	if _, err := aes.NewCipher(key); err != nil {
		return nil, err
	}
	return &Session{SessionID: sessionID, key: key}, nil
}

func (s *Session) aad() []byte {
	return []byte(AADInfo + "/" + s.SessionID)
}

// Seal produces the JSON envelope {"nonce","ct"} used by the HTTP transport.
func (s *Session) Seal(plain []byte) (map[string]string, error) {
	gcm, err := s.gcm()
	if err != nil {
		return nil, err
	}
	nonce := make([]byte, NonceLen)
	if _, err := rand.Read(nonce); err != nil {
		return nil, err
	}
	ct := gcm.Seal(nil, nonce, plain, s.aad())
	return map[string]string{"nonce": B64(nonce), "ct": B64(ct)}, nil
}

// Open decrypts an HTTP envelope.
func (s *Session) Open(nonceB64, ctB64 string) ([]byte, error) {
	nonce, err := B64D(nonceB64)
	if err != nil {
		return nil, err
	}
	if len(nonce) != NonceLen {
		return nil, ErrLen
	}
	ct, err := B64D(ctB64)
	if err != nil {
		return nil, err
	}
	gcm, err := s.gcm()
	if err != nil {
		return nil, err
	}
	pt, err := gcm.Open(nil, nonce, ct, s.aad())
	if err != nil {
		return nil, ErrDecrypt
	}
	return pt, nil
}

// SealCompact produces FRAME_CHECKIN | nonce | ct for the DNS transport.
func (s *Session) SealCompact(plain []byte) ([]byte, error) {
	gcm, err := s.gcm()
	if err != nil {
		return nil, err
	}
	nonce := make([]byte, NonceLen)
	if _, err := rand.Read(nonce); err != nil {
		return nil, err
	}
	ct := gcm.Seal(nil, nonce, plain, s.aad())
	out := make([]byte, 0, 1+NonceLen+len(ct))
	out = append(out, FrameCheckin)
	out = append(out, nonce...)
	out = append(out, ct...)
	return out, nil
}

// OpenCompact decrypts a DNS checkin frame.
func (s *Session) OpenCompact(framed []byte) ([]byte, error) {
	if len(framed) < 1+NonceLen+16 || framed[0] != FrameCheckin {
		return nil, ErrFrame
	}
	nonce := framed[1 : 1+NonceLen]
	ct := framed[1+NonceLen:]
	gcm, err := s.gcm()
	if err != nil {
		return nil, err
	}
	pt, err := gcm.Open(nil, nonce, ct, s.aad())
	if err != nil {
		return nil, ErrDecrypt
	}
	return pt, nil
}

func (s *Session) gcm() (cipher.AEAD, error) {
	block, err := aes.NewCipher(s.key)
	if err != nil {
		return nil, err
	}
	return cipher.NewGCM(block)
}
